File: sdks/javascript/test/didcomm.test.ts
import { describe, expect, it } from "vitest";
import { ATAPClient } from "../src/client.js";
import { generateKeypair } from "../src/crypto.js";
import {
  createMockFetch,
  SAMPLE_DIDCOMM_MESSAGE,
  SAMPLE_TOKEN_RESPONSE,
} from "./helpers.js";

async function createAuthedClient(mockFetch: ReturnType<typeof createMockFetch>) {
  const { privateKey } = await generateKeypair();
  return new ATAPClient({
    baseUrl: "http://localhost:8080",
    did: "did:web:localhost%3A8080:agent:test",
    signingKey: privateKey,
    clientSecret: "atap_secret",
    fetch: mockFetch,
  });
}

describe("DIDCommAPI", () => {
  describe("send", () => {
    it("sends a DIDComm message", async () => {
      const mockFetch = createMockFetch([
        { status: 202, body: { id: "msg_new", status: "queued" } },
      ]);
      const client = new ATAPClient({
        baseUrl: "http://localhost:8080",
        fetch: mockFetch,
      });

      const jweBytes = new TextEncoder().encode(
        JSON.stringify({ protected: "header", ciphertext: "data" }),
      );
      const result = await client.didcomm.send(jweBytes);

      expect(result.id).toBe("msg_new");
      expect(result.status).toBe("queued");

      const headers = mockFetch.calls[0].init.headers as Record<string, string>;
      expect(headers["Content-Type"]).toBe(
        "application/didcomm-encrypted+json",
      );
    });
  });

  describe("inbox", () => {
    it("retrieves inbox messages", async () => {
      const mockFetch = createMockFetch([
        { status: 200, body: SAMPLE_TOKEN_RESPONSE },
        {
          status: 200,
          body: {
            messages: [SAMPLE_DIDCOMM_MESSAGE, SAMPLE_DIDCOMM_MESSAGE],
            count: 2,
          },
        },
      ]);
      const client = await createAuthedClient(mockFetch);

      const inbox = await client.didcomm.inbox();

      expect(inbox.count).toBe(2);
      expect(inbox.messages).toHaveLength(2);
      expect(inbox.messages[0].id).toBe("msg_abc123");
      expect(inbox.messages[0].senderDid).toContain("did:web:");
      expect(inbox.messages[0].messageType).toBe(
        "https://atap.dev/protocol/signal",
      );
      expect(inbox.messages[0].payload).toBe('{"hello":"world"}');
    });

    it("respects limit parameter", async () => {
      const mockFetch = createMockFetch([
        { status: 200, body: SAMPLE_TOKEN_RESPONSE },
        { status: 200, body: { messages: [], count: 0 } },
      ]);
      const client = await createAuthedClient(mockFetch);

      await client.didcomm.inbox({ limit: 10 });

      expect(mockFetch.calls[1].url).toContain("limit=10");
    });

    it("caps limit at 100", async () => {
      const mockFetch = createMockFetch([
        { status: 200, body: SAMPLE_TOKEN_RESPONSE },
        { status: 200, body: { messages: [], count: 0 } },
      ]);
      const client = await createAuthedClient(mockFetch);

      await client.didcomm.inbox({ limit: 200 });

      expect(mockFetch.calls[1].url).toContain("limit=100");
    });

    it("defaults limit to 50", async () => {
      const mockFetch = createMockFetch([
        { status: 200, body: SAMPLE_TOKEN_RESPONSE },
        { status: 200, body: { messages: [], count: 0 } },
      ]);
      const client = await createAuthedClient(mockFetch);

      await client.didcomm.inbox();

      expect(mockFetch.calls[1].url).toContain("limit=50");
    });

    it("handles empty inbox", async () => {
      const mockFetch = createMockFetch([
        { status: 200, body: SAMPLE_TOKEN_RESPONSE },
        { status: 200, body: { messages: [], count: 0 } },
      ]);
      const client = await createAuthedClient(mockFetch);

      const inbox = await client.didcomm.inbox();
      expect(inbox.messages).toHaveLength(0);
      expect(inbox.count).toBe(0);
    });
  });
});

File: sdks/go/didcomm_test.go
package atap

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestDIDCommAPI_Send(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/didcomm" {
			w.WriteHeader(404)
			return
		}
		if r.Header.Get("Content-Type") != "application/didcomm-encrypted+json" {
			t.Errorf("Content-Type = %q", r.Header.Get("Content-Type"))
		}
		body, _ := io.ReadAll(r.Body)
		if string(body) != `{"test":"jwe"}` {
			t.Errorf("body = %q", string(body))
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(202)
		_ = json.NewEncoder(w).Encode(map[string]interface{}{
			"id":     "msg_123",
			"status": "queued",
		})
	}))
	defer server.Close()

	client, _ := NewClient(WithBaseURL(server.URL))
	result, err := client.DIDComm.Send(context.Background(), []byte(`{"test":"jwe"}`))
	if err != nil {
		t.Fatalf("Send: %v", err)
	}
	if result["id"] != "msg_123" {
		t.Errorf("id = %v", result["id"])
	}
	if result["status"] != "queued" {
		t.Errorf("status = %v", result["status"])
	}
}

func TestDIDCommAPI_Inbox(t *testing.T) {
	routes := map[string]mockRoute{
		"GET /v1/didcomm/inbox": {
			Status: 200,
			Body: map[string]interface{}{
				"count": 2,
				"messages": []interface{}{
					map[string]interface{}{
						"id":           "msg_1",
						"sender_did":   "did:web:localhost:agent:a1",
						"message_type": "text",
						"payload":      "hello",
						"created_at":   "2024-01-01T00:00:00Z",
					},
					map[string]interface{}{
						"id":           "msg_2",
						"sender_did":   "did:web:localhost:agent:a2",
						"message_type": "request",
						"payload":      "data",
					},
				},
			},
		},
	}
	server := newOAuthMockServer(t, routes)
	defer server.Close()

	client := newTestClient(t, server)
	inbox, err := client.DIDComm.Inbox(context.Background(), 50)
	if err != nil {
		t.Fatalf("Inbox: %v", err)
	}
	if inbox.Count != 2 {
		t.Errorf("Count = %d", inbox.Count)
	}
	if len(inbox.Messages) != 2 {
		t.Fatalf("Messages count = %d", len(inbox.Messages))
	}
	if inbox.Messages[0].ID != "msg_1" {
		t.Errorf("first message ID = %q", inbox.Messages[0].ID)
	}
	if inbox.Messages[0].SenderDID != "did:web:localhost:agent:a1" {
		t.Errorf("first message SenderDID = %q", inbox.Messages[0].SenderDID)
	}
}

func TestDIDCommAPI_Inbox_Empty(t *testing.T) {
	routes := map[string]mockRoute{
		"GET /v1/didcomm/inbox": {
			Status: 200,
			Body: map[string]interface{}{
				"count":    0,
				"messages": []interface{}{},
			},
		},
	}
	server := newOAuthMockServer(t, routes)
	defer server.Close()

	client := newTestClient(t, server)
	inbox, err := client.DIDComm.Inbox(context.Background(), 0) // default limit
	if err != nil {
		t.Fatalf("Inbox: %v", err)
	}
	if inbox.Count != 0 {
		t.Errorf("Count = %d", inbox.Count)
	}
	if len(inbox.Messages) != 0 {
		t.Errorf("Messages count = %d", len(inbox.Messages))
	}
}

func TestDIDCommAPI_Inbox_LimitClamped(t *testing.T) {
	routes := map[string]mockRoute{
		"GET /v1/didcomm/inbox": {
			Status: 200,
			Body: map[string]interface{}{
				"count":    0,
				"messages": []interface{}{},
			},
		},
	}
	server := newOAuthMockServer(t, routes)
	defer server.Close()

	client := newTestClient(t, server)
	// limit > 100 should be clamped
	_, err := client.DIDComm.Inbox(context.Background(), 200)
	if err != nil {
		t.Fatalf("Inbox: %v", err)
	}
}
